### PR TITLE
Add tracing support for optional Device and Layout

### PR DESCRIPTION
**Pull request by izdeby**
This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

--------------
In this PR: 
Add tracing support for optional Device and Layout types.

--------------

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30983 Updated codegen to remove TensorOptions from generated code
* #30982 Prepare templates
* #30981 Updated DispatchKeyExtractor to expect TensorOptions
* #30980 Added C++ API test
* **#30979 Add tracing support for optional Device and Layout**

Differential Revision: [D18912685](https://our.internmc.facebook.com/intern/diff/D18912685)

### CODE REVIEW

**Comment by gchanan** [comment]
Someone from JIT (e.g. @eellison should review this).

File: torch/csrc/jit/tracer.cpp
Comment (eellison) [inline torch/csrc/jit/tracer.cpp]: It would be nice if you factored the 8 copy-pasta's of this three-line block into it's own function (i am one of those copy-paster's...)

**Review by eellison — APPROVED** [review]
LGTM. I don't know how exactly you plan on landing this stack; could you not land this PR without adding tests for these changes ? If it's in an upcoming PR that's fine.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**
